third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -1742,19 +1745,31 @@ impl Loader {
                 .into_iter()
                 .unzip();
 
+        // Some types can have fields which have custom layouts at serialization or
+        // deserialization times. If so, they must be tagged.
+        let maybe_layout_tag = self.native_types.get_layout_tag(name);
         let has_identifier_mappings =
-            maybe_mapping.is_some() || field_has_identifier_mappings.into_iter().any(|b| b);
+            maybe_layout_tag.is_some() || field_has_identifier_mappings.into_iter().any(|b| b);
 
         let field_node_count = *count - count_before;
-        let layout = if Some(IdentifierMappingKind::DerivedString) == maybe_mapping {
+
+        // Construct struct layout.
+        // TODO[agg_v2](cleanup): Once we have a native value impl, we can also
+        //   construct "native", i.e. tagged layouts for these native fields, such
+        //   as MoveTypeLayout::Native(u64) which are fast to propagate. This can
+        //   allow us to shift matching logic below into adapter.
+        let layout = if Some(LayoutTag::IdentifierMapping(
```

Comment:
seems like you are generalizing for the loader, but we still have aggregator-specific special handling there - i.e. for some replace a first field in a struct

maybe we should see how to express that generically? 

maybe: `tag_field_index: Option<usize>` ?